src/QLPlan.actor.cpp

```diff
@@ -687,38 +687,58 @@ ACTOR static Future<Void> doNonIsolatedRO(PlanCheckpoint* outerCheckpoint,
                                           Reference<MetadataManager> mm) {
 	if (!dtr)
 		dtr = self->newTransaction();
-	state double startt = now();
 	state Reference<PlanCheckpoint> innerCheckpoint(new PlanCheckpoint);
 	state int nTransactions = 1;
 	state int64_t nResults = 0;
 	state FlowLock* outerLock = outerCheckpoint->getDocumentFinishedLock();
+	state Deque<std::pair<Reference<ScanReturnedContext>, Future<Void>>> bufferedDocs;
+
 	try {
 		state uint64_t metadataVersion = wait(cx->bindCollectionContext(dtr)->getMetadataVersion());
 		loop {
 			state FutureStream<Reference<ScanReturnedContext>> docs = subPlan->execute(innerCheckpoint.getPtr(), dtr);
 			state FlowLock* innerLock = innerCheckpoint->getDocumentFinishedLock();
 			state bool first = true;
-			state Future<Void> timeout = delay(3.0);
+			state Future<Void> timeout = delay(3.0, TaskMaxPriority);
+			state bool finished = false;
 
-			loop choose {
-				when(state Reference<ScanReturnedContext> doc = waitNext(docs)) {
-					// throws end_of_stream when totally finished
-					Void _ = wait(outerLock->take());
```

Comment:
Since we no longer use `outerLock` to do flow control, now there may be cases where:

1. Some user does lots of large queries with `numberToReturn` set and forgot to drain the cursors somehow;
2. At the same time our house keep coding failed to catch up with that and didn't remove stale cursors quickly enough,

We may then see memory pressure go rocket high. But I think that's pretty rare and much acceptable compared with the timeout error this PR tries to fix.

Reply:
I am blind.